src/Storages/ExternalTable/ClickHouse/ClickHouse.cpp

```diff
@@ -0,0 +1,125 @@
+#include <Storages/ExternalTable/ClickHouse/ClickHouse.h>
+#include <Storages/ExternalTable/ClickHouse/ClickHouseSink.h>
+#include <Storages/ExternalTable/ClickHouse/ClickHouseSource.h>
+#include <Storages/ColumnsDescription.h>
+#include "Storages/ExternalTable/ExternalTableFactory.h"
+
+namespace DB
+{
+
+namespace ExternalTable
+{
+
+ClickHouse::ClickHouse(const String & name, ExternalTableSettingsPtr settings)
+    : database(settings->database.value)
+    , table(settings->table.changed ? settings->table.value : name)
+    , logger(&Poco::Logger::get("ExternalTable-ClickHouse-" + table))
+{
+    assert(settings->type.value == "clickhouse");
+
+    auto addr = settings->address.value;
+    auto pos = addr.find_first_of(':');
+    if (pos == String::npos)
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "Invalid ClickHouse address, expected format '<host>:<port>'");
+    auto host = addr.substr(0, pos);
+    auto port = std::stoi(addr.substr(pos + 1));
+    if (!port)
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "Invalid port in ClickHouse address");
+
+    connection_params.host = host;
+    connection_params.port = port;
+    connection_params.user = settings->user.value;
+    connection_params.password = settings->password.value;
+    connection_params.default_database = settings->database.value;
+    connection_params.security = settings->secure.value ? Protocol::Secure::Enable : Protocol::Secure::Disable;
+    /// TODO read from settings
+    connection_params.timeouts = {
+        /*connection_timeout_=*/ 1 * 60 * 1'000'000,
+        /*send_timeout_=*/ 1 * 60 * 1'000'000,
+        /*receive_timeout_=*/ 1 * 60 * 1'000'000,
+        /*tcp_keep_alive_timeout_=*/ 5 * 60 * 1'000'000
+    };
+}
+
+void ClickHouse::startup()
+{
+    LOG_INFO(logger, "startup");
+}
+
+Pipe ClickHouse::read(
+    const Names & column_names,
+    const StorageSnapshotPtr & storage_snapshot,
+    SelectQueryInfo &  /*query_info*/,
+    ContextPtr context,
+    QueryProcessingStage::Enum processed_stage,
+    size_t  /*max_block_size*/,
+    size_t /*num_streams*/)
+{
+    auto header = storage_snapshot->getSampleBlockForColumns(column_names);
+    auto client = std::make_unique<ClickHouseClient>(connection_params, logger);
+    auto source = std::make_shared<ClickHouseSource>(database, table, std::move(header), std::move(client), processed_stage, context, logger);
+    return Pipe(std::move(source));
+}
+
+SinkToStoragePtr ClickHouse::write(const ASTPtr &  /*query*/, const StorageMetadataPtr & metadata_snapshot, ContextPtr  context)
+{
+    return std::make_shared<ClickHouseSink>(database, table, metadata_snapshot->getSampleBlock(), connection_params, context, logger);
+}
+
+ColumnsDescription ClickHouse::getTableStructure()
+{
+    ColumnsDescription ret {};
+
+    ClickHouseClient client {connection_params, logger};
+    auto query = "DESCRIBE TABLE " + (database.empty() ? "" : backQuoteIfNeed(database) + ".") + backQuoteIfNeed(table);
```

Comment:
we can try `auto query = fmt::format("DESCRIBE TABLE {}...")` which may be a bit more readable.